app/src/test/java/org/mozilla/fenix/components/toolbar/CFRPresenterTest.kt

```diff
@@ -0,0 +1,324 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.fenix.components.toolbar
+
+import android.content.Context
+import android.view.View
+import androidx.compose.ui.unit.dp
+import io.mockk.Runs
+import io.mockk.every
+import io.mockk.just
+import io.mockk.mockk
+import io.mockk.spyk
+import io.mockk.verify
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.cancel
+import mozilla.components.browser.state.action.ContentAction
+import mozilla.components.browser.state.state.BrowserState
+import mozilla.components.browser.state.state.createCustomTab
+import mozilla.components.browser.state.state.createTab
+import mozilla.components.browser.state.store.BrowserStore
+import mozilla.components.browser.toolbar.BrowserToolbar
+import mozilla.components.support.test.ext.joinBlocking
+import mozilla.components.support.test.robolectric.testContext
+import mozilla.components.support.test.rule.MainCoroutineRule
+import org.junit.Assert.assertEquals
+import org.junit.Assert.assertFalse
+import org.junit.Assert.assertNotNull
+import org.junit.Assert.assertNull
+import org.junit.Assert.assertTrue
+import org.junit.Rule
+import org.junit.Test
+import org.mozilla.fenix.R
+import org.mozilla.fenix.compose.cfr.CFRPopup
+import org.mozilla.fenix.utils.Settings
+import mozilla.telemetry.glean.testing.GleanTestRule
+import org.junit.runner.RunWith
+import org.mozilla.fenix.GleanMetrics.TrackingProtection
+import org.mozilla.fenix.helpers.FenixRobolectricTestRunner
+
+@RunWith(FenixRobolectricTestRunner::class)
+class CFRPresenterTest {
+    @get:Rule
+    val coroutinesTestRule = MainCoroutineRule()
+    @get:Rule
+    val gleanTestRule = GleanTestRule(testContext)
+
+    @Test
+    fun `GIVEN the TCP CFR should be shown for a custom tab WHEN the custom tab is fully loaded THEN the TCP CFR is shown`() {
+        val customTab = createCustomTab(url = "")
+        val browserStore = BrowserStore(
```

Comment:
Throughout the test we create `browserStore` and `presenter` multiple times. Given it takes a few lines of code, may be it's worth creating a function (with some parameters in case we need to change those like `DefaultBrowserToolbarMenuControllerTest.createController()`?

Reply:
Looked into this but the arguments are different in most cases so if having to most often pass different parameters the improvements are minimal. Eg: after refactoring

```kotlin
        val presenter = spyk(
            CFRPresenter(
                context = mockk(),
                browserStore = browserStore,
                settings = mockk {
                    every { shouldShowTotalCookieProtectionCFR } returns true
                },
                toolbar = mockk(),
                sessionId = customTab.id
            )
        ) {
            every { showTcpCfr() } just Runs
        }
```

I would get:
```kotlin
        val presenter = spyk(
            createPresenter(
                browserStore = browserStore,
                settings = mockk {
                    every { shouldShowTotalCookieProtectionCFR } returns true
                },
                sessionId = customTab.id
            )
        ) {
            every { showTcpCfr() } just Runs
        }
```

Reply:
It is not the only way to do it, we could also use the default parameters to set the default values to parameters the repeat themselves from test to test. I played around with the tests, and we could reduce the file size by ~60 lines we put the repeated code into constructors, and the rest would like this
```
val customTab = createCustomTab(url = "")
val browserStore = createBrowser(customTab = customTab)
val presenter = createSpyPresenter(browserStore = browserStore, sessionId = customTab.id)
```

Reply:
this is what I went for, and I checked that tests were working
```
    private fun createSpyPresenter(
        context: Context = mockk(),
        anchor: View = mockk(),
        browserStore: BrowserStore = mockk(),
        settings: Settings = mockk { every { shouldShowTotalCookieProtectionCFR } returns true },
        toolbar: BrowserToolbar = mockk(),
        sessionId: String? = null
    ) = spyk(createPresenter(context, anchor, browserStore, settings, toolbar, sessionId)) {
        every { showTcpCfr() } just Runs
    }

    private fun createPresenter(
        context: Context = mockk { every { getString(R.string.tcp_cfr_message) } returns "Test" },
        anchor: View = mockk(),
        browserStore: BrowserStore = mockk(),
        settings: Settings = mockk { every { shouldShowTotalCookieProtectionCFR } returns true },
        toolbar: BrowserToolbar = mockk {
            every { findViewById<View>(R.id.mozac_browser_toolbar_security_indicator) } returns anchor
        },
        sessionId: String? = null
    ) = CFRPresenter(
        context = context,
        browserStore = browserStore,
        settings = settings,
        toolbar = toolbar,
        sessionId = sessionId
    )

    private fun createBrowser(
        normalTab: TabSessionState? = null,
        customTab: CustomTabSessionState? = null,
        selectedTabId: String? = null
    ) = BrowserStore(
        initialState = BrowserState(
            tabs = if (normalTab != null) listOf(normalTab) else listOf(),
            selectedTabId = selectedTabId,
            customTabs = if (customTab != null) listOf(customTab) else listOf()
        )
    )
```

Reply:
Done.